en/FeedHeader.md

```diff
@@ -7,4 +7,4 @@
 | gtfs_realtime_version | Current version is "2.0".  All GTFS Realtime feeds should be "2.0" or higher, as early version of GTFS Realtime did not require all fields needed to represent various transit situations adequately. |
 | incrementality | | 
 | timestamp | This timestamp should not decrease between two sequential feed iterations. |
-|  | This timestamp value should always change if the feed contents change - the feed contents should not change without updating the header `timestamp`.<br>*Common mistakes* - If there are multiple instances of GTFS Realtime feed behind a load balancer, each instance may be pulling information from the realtime data source and publishing it to consumers slightly out of sync. If a GTFS Realtime consumer makes two back-to-back requests, and each request is served by a different GTFS Realtime feed instance, the same feed contents could potentially be returned to the consumer with different timestamps.<br>*Possible solution* - Configure the load balancer for "sticky routes", so that the consumer always receives the GTFS Realtime feed contents from the same GTFS Realtime instance. |
+|  | This timestamp value should always change if the feed contents change - the feed contents should not change without updating the header `timestamp`.<br>*Common mistakes* - If there are multiple instances of GTFS Realtime feed behind a load balancer, each instance may be pulling information from the realtime data source and publishing it to consumers slightly out of sync. If a GTFS Realtime consumer makes two back-to-back requests, and each request is served by a different GTFS Realtime feed instance, the same feed contents could potentially be returned to the consumer with different timestamps.<br>*Possible solution* - Producers should provide `Last-Modified` or `Etag` HTTP headers, and consumers should pass a `If-Modified-Since` or `If-None-Match` HTTP header to avoid receiving stale data. |
```

Comment:
Good suggestion! IIRC in Tampa this wasn't possible though for some reason related to internal infrastructure, so the "sticky routes" solution was used instead. Maybe we should list them both as "possible solutions", with the HTTP header solution listed first?

Reply:
my 2 cents: To be honest, I don't see why this (bad) internal design of a load balancing system should lead the GTFS-RT Best Practices to disregard well-known and commonly used HTTP semantics. 😉

Reply:
If there is a consensus that sticky routes here is "bad design" and shouldn't be used in practice, and there isn't anything blocking use of HTTP headers in the majority of deployments, I don't object to removing the mention of sticky routes here.

Reply:
Okay, my post is a bit misleading, sorry. I don't see a reason not to use sticky routes internally, but no matter how the load balancing system is built, doing requests from the outside, it should act consistently: I shouldn't *ever* receive an older GTFS-RT feed on a subsequent request, right?

Or to rephrase it again: In this case, using "sticky routes" is a bad and insufficient solution to a *different* problem: one of internal consistency in the publisher's system. What if I request the GTFS-RT dump from 2 different IP addresses but process it in one place? In this case my system is *one* consumer but the load balancer has no way to guarantee a consistent behaviour.

Reply:
>I shouldn't ever receive an older GTFS-RT feed on a subsequent request, right?

That's ideal, yes.

>What if I request the GTFS-RT dump from 2 different IP addresses but process it in one place? In this case my system is one consumer but the load balancer has no way to guarantee a consistent behaviour.

Yes, this is true. 

So maybe a better option is to further elaborate here on the desired behavior (consistency from the consumer's perspective) and how both of these options address (or don't address) the problem? 

In my experience often agencies don't have the in-house expertise to understand the pros/cons of their decisions on consumers.

Reply:
> So maybe a better option is to further elaborate here on the desired behavior (consistency from the consumer's perspective) and how both of these options address (or don't address) the problem?
> 
> In my experience often agencies don't have the in-house expertise to understand the pros/cons of their decisions on consumers.

I agree. But while using `Last-Modified` & `ETag` headers requires technical expertise (configuring a web server like nginx properly), it is a lot harder to get wrong than configuring a load balancer to use sticky routes *only*.

We could rephrase it to say that:

- `Last-Modified` & `ETag` headers are recommended. In combination with [conditional requests](https://developer.mozilla.org/en-US/docs/Web/HTTP/Conditional_requests), they are the intended way of *correctly* (and efficiently) specifying that or checking if a resource has changed.
- *If* agencies want to use a load balancer, they should make sure that all servers behind it have a consistent state, e.g. by synchronising the generated GTFS-RT data.

---

Related to the topic:

Configuring a single caching (reverse) proxy to

- periodically re-fetch the original/actual GTFS-RT feed when there's a request (e.g. every few seconds), *and*
- serve *all* requests from the cache

is a semantically correct and probably the most efficient setup. AFAIK most popular HTTP (reverse) proxies can be configured to do that, and having >1 GTFS-RT-generating instances shouldn't be necessary anymore, as it is guaranteed not to receive spikes of requests.

Reply:
On second thought, `ETag` doesn't actually help in this case since you could flip between two cache endpoints with different data. I'll work on re-phrasing.